### PR TITLE
Implement UnwrapArrayMap and UnwrapArrayFilter mutators

### LINKED ISSUE #490: Function call unwrapping mutators
[A fine example from our Slack channel](https://symfony-devs.slack.com/archives/C78M7A4BY/p1537651009000100) by @borNfreee:

```diff
- $lowercasedStrings = array_map(['L', 'O', 'L'], function ($v) {
-    return strtolower($v);
- });
+ $lowercasedStrings = ['L', 'O', 'L'];
```
And just as such for `array_filter`. There could be more similar mutators which all will be looking like:

```diff
- $output = standardFunction($input, $foo, $bar);
+ $output = $input;
```

### PR DESCRIPTION
**Pull request by akondas**
This PR:

- [x] Create `AbstractUnwrapMutator`
- [x] Implement `UnwrapArrayFilter` and `UnwrapArrayMap` mutators
- [x] Covered by tests
- [ ] Doc PR: see https://github.com/infection/site/pulls for inspiration

Inspiration in #490 

Fixes #490

### CODE REVIEW

**Comment by sanmai** [comment]
These seem like fine examples for our new generating mutatiors:

https://github.com/infection/infection/blob/68f70ad62ffd14af62781d989e4230c45c235010/src/Mutator/Arithmetic/RoundingFamily.php#L42-L48

That way you can also cover `array_flip()`, `array_reverse()`, `array_replace()`, and others, without much extra effort. Sure, tests will be longer, but that's unavoidable.

File: src/Mutator/Unwrap/AbstractUnwrapMutator.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    abstract protected function getFunctionName(): string;
+
+    abstract protected function getParameterIndex(): int;
+
+    /**
+     * Replaces "$a = function(arg1, arg2);" with "$a = arg1;"
+     *
+     * @return Node\Param;
+     */
+    public function mutate(Node $node)
```
Comment (sanmai) [inline src/Mutator/Unwrap/AbstractUnwrapMutator.php]:
```suggestion
    final public function mutate(Node $node)
```

File: tests/Mutator/Unwrap/UnwrapArrayFilterTest.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+$a = ['A', 1, 'C'];
+PHP
+        ];
+
+        yield [
+            <<<'PHP'
+<?php
+
+$a = array_filter(['A', 1, 'C'], function($var): bool {
```
Comment (borNfreee) [inline tests/Mutator/Unwrap/UnwrapArrayFilterTest.php]: could you please add more tests (like parameter as a variable, function with lower/upper cased characters and so on.)

See the list of cases in https://github.com/infection/infection/blob/master/tests/Mutator/Regex/PregQuoteTest.php

same for `array_map`

File: src/Mutator/Unwrap/AbstractUnwrapMutator.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        return $node->args[$this->getParameterIndex()];
+    }
+
+    protected function mutatesNode(Node $node): bool
```
[the remark this answers is not in the document]
Comment (akondas) [reply src/Mutator/Unwrap/AbstractUnwrapMutator.php]: Definietly :)

File: tests/Mutator/Unwrap/UnwrapArrayFilterTest.php (hunk 2 above)
Comment (akondas) [reply tests/Mutator/Unwrap/UnwrapArrayFilterTest.php]: Nice catch, wrongly capitalized was the problem :)

Comment (sanmai) [reply tests/Mutator/Unwrap/UnwrapArrayFilterTest.php]: There's `$node->name->toLowerString()` of use.

File: tests/Mutator/Unwrap/UnwrapArrayFilterTest.php [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+<?php
+
+$a = ['A', 1, 'C'];
+if ($a === $a) {
+    return true;
+}
+PHP
+        ];
+
+        yield 'It mutates correctly when array_map is wrongly capitalized' => [
```
Comment (sanmai) [inline tests/Mutator/Unwrap/UnwrapArrayFilterTest.php]:
```suggestion
        yield 'It mutates correctly when array_filter is wrongly capitalized' => [
```

File: tests/Mutator/Unwrap/UnwrapArrayFilterTest.php [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+PHP
+            ,
+            <<<'PHP'
+<?php
+
+$a = ['A', 1, 'C'];
+PHP
+        ];
+
+        yield 'It mutates correctly when array_map uses another function as input' => [
```
Comment (sanmai) [inline tests/Mutator/Unwrap/UnwrapArrayFilterTest.php]:
```suggestion
        yield 'It mutates correctly when array_filter uses another function as input' => [
```

**Comment by akondas** [comment]
Can we restart CI jobs? `Failed loading C:\tools\php\ext\xdebug.dll` it looks like an unreliable error.

**Comment by borNfreee** [comment]
Thank you @akondas, nice work.

The build is green now

**Comment by borNfreee** [comment]
@akondas can I ask you to update documentation, please? See https://github.com/infection/site/issues/70

**Comment by borNfreee** [comment]
I just realized that with this new base abstract class, we can implement not only array functions mutators as mentioned [here](https://github.com/infection/infection/issues/490#issuecomment-426157145), but for any other function, for example

```diff
- return strtolower($string);
+ return $string;
```

👍

**Comment by sanmai** [comment]
Yeah, and things like:
```
- return array_reduce($array, $func, $initial);
+ return $initial;
```

Should we make a list within an issue with ideas?

**Comment by borNfreee** [comment]
good idea, done https://github.com/infection/infection/issues/514

File: src/Mutator/Util/MutatorProfile.php [hunk 6]
```diff
@@ -159,6 +160,11 @@
         Mutator\Cast\CastString::class,
     ];
 
+    public const UNWRAP = [
+        Mutator\Unwrap\UnwrapArrayMap::class,
+        Mutator\Unwrap\UnwrapArrayFilter::class,
```
Comment (localheinz) [inline src/Mutator/Util/MutatorProfile.php]: @akondas @borNfreee @sanmai 

Is the order important here?

If not, maybe we can sort alphabetically?

File: src/Mutator/Util/MutatorProfile.php [hunk 7]
```diff
@@ -278,5 +284,9 @@
         'CastInt' => Mutator\Cast\CastInt::class,
         'CastObject' => Mutator\Cast\CastObject::class,
         'CastString' => Mutator\Cast\CastString::class,
+
+        // Unwrap
+        'UnwrapArrayMap' => Mutator\Unwrap\UnwrapArrayMap::class,
+        'UnwrapArrayFilter' => Mutator\Unwrap\UnwrapArrayFilter::class,
```
Comment (localheinz) [inline src/Mutator/Util/MutatorProfile.php]: @akondas @borNfreee @sanmai 

Is the order important here?

If not, maybe we can sort alphabetically?

File: src/Mutator/Util/MutatorProfile.php (hunk 6 above)
Comment (sanmai) [reply src/Mutator/Util/MutatorProfile.php]: Same order [as in the manual](http://php.net/manual/en/book.array.php) should work best, IMO.

**Pull request merged**